lxd/container_lxc.go

```diff
@@ -6883,13 +6887,139 @@ func (c *containerLXC) insertProxyDevice(devName string, m types.Device) error {
+		// address otherwise we could mess with other containers.
+		if ip != "" && IPv4Addr == "" && (address == ip || address == "0.0.0.0") {
+			IPv4Addr = ip
+		}
+
+		ip = m["ipv6.address"]
+		if ip != "" && IPv6Addr == "" && (address == ip || address == "::") {
+			IPv6Addr = ip
+		}
+	}
+
+	if IPv4Addr == "" && IPv6Addr == "" {
+		return errors.New("NIC IP doesn't match proxy target IP")
+	}
+
+	iptablesComment := fmt.Sprintf("%s (%s)", c.Name(), proxy)
+
+	revert := true
+	defer func() {
+		if revert {
+			if IPv4Addr != "" {
+				containerIptablesClear("ipv4", iptablesComment, "nat")
+			}
+
+			if IPv6Addr != "" {
+				containerIptablesClear("ipv6", iptablesComment, "nat")
+			}
+		}
+	}()
+
+	for i, lAddr := range listenAddr.addr {
+		address, port, err := net.SplitHostPort(lAddr)
+		if err != nil {
+			return err
+		}
+		var cPort string
+		if len(connectAddr.addr) == 1 {
+			_, cPort, _ = net.SplitHostPort(connectAddr.addr[0])
+		} else {
+			_, cPort, _ = net.SplitHostPort(connectAddr.addr[i])
+		}
+
+		if IPv4Addr != "" {
+			// outbound <-> container
+			err := containerIptablesPrepend("ipv4", iptablesComment, "nat",
+				"PREROUTING", "-p", listenAddr.connType, "--destination",
+				address, "--dport", port, "-j", "DNAT",
+				"--to-destination", fmt.Sprintf("%s:%s", IPv4Addr, cPort))
+			if err != nil {
+				return err
+			}
+
+			// host <-> container
+			err = containerIptablesPrepend("ipv4", iptablesComment, "nat",
+				"OUTPUT", "-p", listenAddr.connType, "--destination",
+				address, "--dport", port, "-j", "DNAT",
+				"--to-destination", fmt.Sprintf("%s:%s", IPv4Addr, cPort))
+			if err != nil {
+				return err
+			}
+		}
+
+		if IPv6Addr != "" {
+			// outbound <-> container
+			err := containerIptablesPrepend("ipv6", iptablesComment, "nat",
+				"PREROUTING", "-p", listenAddr.connType, "--destination",
+				address, "--dport", port, "-j", "DNAT",
+				"--to-destination", fmt.Sprintf("[%s]:%s", IPv6Addr, cPort))
+			if err != nil {
+				return err
+			}
+
+			// host <-> container
+			err = containerIptablesPrepend("ipv6", iptablesComment, "nat",
+				"OUTPUT", "-p", listenAddr.connType, "--destination",
+				address, "--dport", port, "-j", "DNAT",
+				"--to-destination", fmt.Sprintf("[%s]:%s", IPv6Addr, cPort))
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	revert = false
+	logger.Info(fmt.Sprintf("Using NAT for proxy device '%s'", proxy))
+	return nil
+}
+
 func (c *containerLXC) removeProxyDevice(devName string) error {
 	if !c.IsRunning() {
 		return fmt.Errorf("Can't remove proxy device from stopped container")
 	}
 
+	// Remove possible iptables entries
+	containerIptablesClear("ipv4", fmt.Sprintf("%s (%s)", c.Name(), devName), "nat")
+	containerIptablesClear("ipv6", fmt.Sprintf("%s (%s)", c.Name(), devName), "nat")
+
 	devFileName := fmt.Sprintf("proxy.%s", devName)
 	devPath := filepath.Join(c.DevicesPath(), devFileName)
+
+	if !shared.PathExists(devPath) {
```

Comment:
Doesn't shared.IsTrue(c.expandedDevices[devName]["nat"]) work? Or does devName refer to something else here?

Reply:
I tried what you suggested, but `c.expandedDevices[devName]` is empty.